### PR TITLE
better performance!

### PR DESCRIPTION
**Pull request by bung87**

### CODE REVIEW

**Comment by dotnetCarpenter** [comment]
@bung87 interesting. I've never heard that changing function declaration to function expression is faster. Why is that? Is it changing `window` to local parameter variable? Do you have any reference?

I've seen that the "use strict" pragma makes code run slightly faster on V8 but never this.

**Comment by bung87** [comment]
@dotnetCarpenter I just read from  a book <High Performance JavaScript> last weekend,most of these chage,just make sure browser support detection only once,detection every time event happend is ineffective,there's two types of change,one of these is detect browser support return the function immediately，the othor is detect browser support when event happend first time rewrite it's self,and at the first time excute self.

**Comment by dotnetCarpenter** [comment]
rewrite itself could be expensive but if it only happens once then I guess it's an optimization. Can you use [`Window.Performance.now()`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) and get some numbers to back up the theory?

**Comment by dotnetCarpenter** [comment]
Also check with "use strict" and without. And maybe *number*`|0`  [to change Numbers to 32bit integers](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators#Bitwise_logical_operators) if you see any. I think there should be some coordinates that could be optimized with that.

**Comment by bung87** [comment]
@dotnetCarpenter  I get different results every time by using window.Performance.now,also I checked  with "use strict" running test.html ,but duration also different,what's the right way to do time profile?

**Comment by dotnetCarpenter** [comment]
>what's the right way to do time profile?

To repeat ;)

Since javascript gets interpreted and compiled (sometimes compiled twice) the same code has different run times depending on how you run it. What you do is, that you setup up a test that you can run on different  versions of the code. That way you know that the *environment* around your code is the same in all tests. You can look into the [unit test](https://github.com/gtramontina/draggable.js/blob/master/test/draggable.test.js) to see how you can write a small program that manipulate `draggable.js`. Now, you don't want to write unit tests, so cut out all the `describe`, `it` and `expect`stuff. We are only interested in the time elapse.
*Hint*: You can use [happen](https://github.com/tmcw/happen) drag elements.

When you got your test program working. It should not be too short. Then you run it enough times, that the profile timings you get from the delta value of `Performance.now()`, stabilize itself around +-20 milliseconds. Try with 10 times for instance. Save all the elapsed times in an array and get the mean value.
E.g.
```js
var mean = [];
for(var i = 0; i < 10; i++) {
  mean.push(doProfiling());
}
console.log( mean ); // print our 10 runs
console.log( mean.reduce(function(prev, curr) { // print the mean value
  return prev + curr;
}, 0) / 10 );
function doProfiling() {
  var start = Performance.now();
  runYourDraggableManipulationTest();
  var end = Performance.now();
  return end - start;
}
```

**Comment by bung87** [comment]
```
  afterEach(function() {
    $(draggableBox).remove();
    end = window.performance.now();
    mean.push(end - start);
  });
```
```
  after(function() {
    console.log( mean.toString());
    console.log( mean.reduce(function(prev, curr) { // print the mean value
      return prev + curr;
    }, 0) /10 );
    // runs after all tests in this block
  });
```

====my version===
1 time
5.341000000044005,4.805999999916821,2.139000000170199,2.957000000151311,2.6669999999739957,2.8250000000298314,3.132000000050539,2.5840000000698637,2.5319999999737774,2.4820000001000153
per: 3.146500000048036
2 time
5.695999999943524,5.549999999857391,2.4880000000848668,3.261999999949694,5.098999999972875,3.4830000001875305,3.71299999983421,3.4349999998539715,3.580000000056316,3.0879999999342544
per: 3.9393999999674634
3 times
5.863999999974112,5.180999999993219,2.7829999999084976,3.74599999986458,6.208999999898879,2.756000000090353,3.560000000106811,3.1440000000202417,4.877000000078624,3.1779999999344
per: 4.129799999986972
4 times
5.523999999695661,5.912999999964086,2.6310000002922607,3.250000000207365,3.2660000001669687,2.507000000150583,10.936000000128843,3.4099999998034036,3.3140000000457803,3.0390000001716544
per 4.3790000000626605
=========the origin===
1 time
5.939000000125816,5.866999999852851,2.5399999999535794,3.776000000016211,4.398999999921216,5.15899999982139,3.541999999924883,3.1510000001162553,3.5080000000107248,3.179000000045562
per: 4.105999999978849
2 times
5.604999999832216,5.580000000009022,2.7549999999791908,3.42899999986912,4.998999999997977,3.512999999884414,3.8839999999709107,3.1520000000000437,5.190999999967971,2.864999999928841
per: 4.097299999943971
3 times
5.541999999877589,5.250999999816486,3.0550000001312583,4.106999999976324,5.202000000053886,2.545999999938431,3.5469999997985724,3.52999999995518,5.395000000135042,3.1549999998787825
per: 4.132999999956155
4 times
5.661000000145577,5.842000000029657,2.6649999999790452,3.40600000026825,5.529000000024098,3.0270000002019515,3.7979999997332925,3.9480000000367,5.011000000195054,2.8919999999743595
per 4.1779000000587985

Am I doing corect?does it help?I ve do this over 10 times,half of my version takes more time than original one,

File: draggable.js [hunk 1]
```diff
@@ -5,54 +5,83 @@
...
-                if (element.addEventListener) {
-                    element.addEventListener(eventName, handler, false);
-                } else if (element.attachEvent) {
-                    element.attachEvent('on' + eventName, handler);
+  var addEventListener = ( function (w) {
+                if (w.addEventListener) {
+                    return function (element, eventName, handler){
+                      element.addEventListener(eventName, handler, false);
+                    }
+                } else if (w.attachEvent) {
```
Comment (dotnetCarpenter) [inline draggable.js]: Unless draggable.js need to support IE8 and below - we can remove the attachEvent stuff. What do you say @gtramontina?

File: draggable.js [hunk 2]
```diff
@@ -5,54 +5,83 @@
...
-  function removeEventListener(element, eventName, handler) {
-      if (element.removeEventListener) {
+            })(window)
+  ,removeEventListener = ( function(w) {
+      if (w.removeEventListener) {
+        return function (element, eventName, handler){
           element.removeEventListener(eventName, handler, false);
-      } else if (element.detachEvent) {
+        }
+      } else if (w.detachEvent) {
```
Comment (dotnetCarpenter) [inline draggable.js]: the same here. Only needed for IE8 and below. btw microsoft stop support for IE8 on January 12, 2016

**Comment by dotnetCarpenter** [comment]
@bung87 I'm afraid that running a bench mark against the test suit will not give any usable result. We need a proper bench mark that reflect a real use-case. It seems that you are profiling the test suit, including the mocha test framework.
As I commented above, we might do with getting away with some code. And the fastest computation is always not to do any. Reading your numbers, it could mean that we need to run the profile test more than 10 times, since the 4th *optimized* code is slower than the slowest of all runs.
But we need a stand alone bench mark test. Not sure if you wrote one as I can't see it. Do you mind committing the bench mark test?

**Comment by bung87** [comment]
So all I need is using happen do the be mark ,if it is i will do this work see if there is any improvement.

**Comment by dotnetCarpenter** [comment]
@Bung You can use `happen` to emulate mouse events. This way you can
program a use case where a user drags some <div>'s around with
`draggable.js`.

**Comment by bung87** [comment]
I simply modify the draggable.test.js posted it on gist,https://gist.github.com/bung87/e18f0de2f3f2ca262cad
run it on the two versions of draggable.js,didnot see much diffrence.

**Comment by gtramontina** [comment]
Hey there! I'm sorry for the silence!
About IE8, we can always drop support for it and release a new major version.
Performance improvements are always welcome! :+1: 

Although I have not used this project lately, I'd like to ask help maintaining it. Would the both of you like to be added as maintainers? :grin:

**Comment by bung87** [comment]
@gtramontina I'd love to,just don't know what's left I can do,Do you have some thoughts of the roadmap?support mobile?or add more feature?

**Comment by dotnetCarpenter** [comment]
Sure. But it will an off and on effort on my part.
@bung87 About your PR, what you're doing is micro optimizations. I think it is always good practice to code with that mind but often it becomes premature optimizations, which is hardly worth it. But since you already did the work, and I don't see any downside to your PR, I think it should be merged.

**Comment by dotnetCarpenter** [comment]
Also, I see that @bung87 made a previous PR to get IE6 support. So my remarks about IE8 and below are void.

**Comment by dotnetCarpenter** [comment]
@bung87 suggestions for improvements.

1. Use travis-ci automatically test PR.
1. Use codecov to see how much the tests cover.
1. Add a general bench mark test for future PRs about performance.

I can easily do all three, unless you really eager for one of them.

**Comment by bung87** [comment]
@dotnetCarpenter Seriously,about your option three,I havn't found a solution to do benmark,since this PR's benchmark can't offer a clear result,do you have any sugestion?

**Pull request merged**